lib/engine-object-file/src/artifact.rs

```diff
@@ -0,0 +1,485 @@
+                byte_buffer[j] = bytes[cur_offset + j];
+            }
+                sp.len = usize::from_ne_bytes(byte_buffer);
+                cur_offset += WORD_SIZE;*/
+
+            let fp = FunctionBodyPtr(mem::transmute(sp));
+            finished_functions.push(fp);
+        }
+
+        let mut signatures: PrimaryMap<_, VMSharedSignatureIndex> = PrimaryMap::new();
+        for i in 0..(sig_map.len()) {
+            if let Some(shared_idx) = sig_map.get(&SignatureIndex::new(i)) {
+                signatures.push(*shared_idx);
+            } else {
+                panic!("Invalid data, missing sig idx; TODO: handle this error");
+            }
+        }
+
+        // read trampolines in order
+        for i in 0..WORD_SIZE {
+            byte_buffer[i] = bytes[cur_offset + i];
+        }
+        cur_offset += WORD_SIZE;
+        let num_function_trampolines = usize::from_ne_bytes(byte_buffer);
+        for i in 0..num_function_trampolines {
+            for j in 0..WORD_SIZE {
+                byte_buffer[j] = bytes[cur_offset + j];
+            }
+            cur_offset += WORD_SIZE;
+            let trampoline_ptr_bytes = usize::from_ne_bytes(byte_buffer);
+            let trampoline = mem::transmute::<usize, VMTrampoline>(trampoline_ptr_bytes);
+
+            let func_type = &metadata.compile_info.module.signatures[SignatureIndex::new(i)];
+
+            engine_inner.add_trampoline(func_type, trampoline);
+            // REVIEW: we can also serialize and read back lengths, do we want to do this?
+            /*for j in 0..WORD_SIZE {
+                byte_buffer[j] = bytes[cur_offset + j];
+            }
+                sp.len = usize::from_ne_bytes(byte_buffer);
+                cur_offset += WORD_SIZE;*/
+        }
+
+        // read dynamic function trampolines in order now...
+        let mut finished_dynamic_function_trampolines = PrimaryMap::new();
+        for i in 0..WORD_SIZE {
+            byte_buffer[i] = bytes[cur_offset + i];
+        }
+        cur_offset += WORD_SIZE;
+        let num_dynamic_trampoline_functions = usize::from_ne_bytes(byte_buffer);
+        for _i in 0..num_dynamic_trampoline_functions {
+            let mut sp = SlicePtr { ptr: 0, len: 0 };
+            for j in 0..WORD_SIZE {
+                byte_buffer[j] = bytes[cur_offset + j];
+            }
+            sp.ptr = usize::from_ne_bytes(byte_buffer);
+            cur_offset += WORD_SIZE;
+
+            // REVIEW: we can also serialize and read back lengths, do we want to do this?
+            /*for j in 0..WORD_SIZE {
+                byte_buffer[j] = bytes[cur_offset + j];
+            }
+                sp.len = usize::from_ne_bytes(byte_buffer);
+                cur_offset += WORD_SIZE;*/
+
+            let fp = FunctionBodyPtr(mem::transmute(sp));
+
+            finished_dynamic_function_trampolines.push(fp);
+        }
+
+        Ok(Self {
+            metadata,
+            // TODO: review
+            module_bytes: vec![],
```

Comment:
I would prefer `None` rather than an empty vec. Thoughts?

Reply:
Yeah, or perhaps we should just store the bytes we deserialized with. That seems like a less leaky abstraction, in general I'd expect a Serialized + Deserialized thing to be identical and for a deserialized thing to be able to be serialized again